以 `pad` 协议登录时, 无法获得来自私聊的音频的下载链接

<!--在下面空白处详细描述你遇到的问题-->
无法获得音频链接
代码
```
for (SingleMessage sm : e.getMessage()) {
    if (sm instanceof OnlineAudio) logger.info("url: {}", ((OnlineAudio) sm).getUrlForDownload());
}
```

#### 日志
```
[18:45:29.371] [net.mamoe.mirai.Bot/ERROR] Exception in unnamed coroutine.
net.mamoe.mirai.event.ExceptionInEventHandlerException: Exception in EventHandler
	at net.mamoe.mirai.internal.event.JvmMethodListenersInternalKt.registerEventHandler$callMethod$invokeWithErrorReport(JvmMethodListenersInternal.kt:147) ~[BotLibV4.jar:?]
	at net.mamoe.mirai.internal.event.JvmMethodListenersInternalKt.access$registerEventHandler$callMethod$invokeWithErrorReport(JvmMethodListenersInternal.kt:1) ~[BotLibV4.jar:?]
	at net.mamoe.mirai.internal.event.JvmMethodListenersInternalKt$registerEventHandler$callMethod$2.invokeSuspend(JvmMethodListenersInternal.kt:154) ~[BotLibV4.jar:?]
	at kotlin.coroutines.jvm.internal.BaseContinuationImpl.resumeWith(ContinuationImpl.kt:33) ~[BotLibV4.jar:?]
	at kotlinx.coroutines.DispatchedTask.run(DispatchedTask.kt:106) ~[BotLibV4.jar:?]
	at kotlinx.coroutines.scheduling.CoroutineScheduler.runSafely(CoroutineScheduler.kt:571) ~[BotLibV4.jar:?]
	at kotlinx.coroutines.scheduling.CoroutineScheduler$Worker.executeTask(CoroutineScheduler.kt:750) ~[BotLibV4.jar:?]
	at kotlinx.coroutines.scheduling.CoroutineScheduler$Worker.runWorker(CoroutineScheduler.kt:678) ~[BotLibV4.jar:?]
	at kotlinx.coroutines.scheduling.CoroutineScheduler$Worker.run(CoroutineScheduler.kt:665) ~[BotLibV4.jar:?]
Caused by: java.lang.reflect.InvocationTargetException
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method) ~[?:1.8.0_282]
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62) ~[?:1.8.0_282]
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43) ~[?:1.8.0_282]
	at java.lang.reflect.Method.invoke(Method.java:498) ~[?:1.8.0_282]
	at net.mamoe.mirai.internal.event.JvmMethodListenersInternalKt.registerEventHandler$callMethod$invokeWithErrorReport(JvmMethodListenersInternal.kt:140) ~[BotLibV4.jar:?]
	... 8 more
Caused by: java.lang.UnsupportedOperationException: Could not fetch URL for audio 99034A2E23A9AEBA711ABA06C8AA8ED7.amr
	at net.mamoe.mirai.internal.message.OnlineAudioImpl.getUrlForDownload(OnlineAudioImpl.kt:118) ~[BotLibV4.jar:?]
	at com.author404e.bot.cmd.CommandParser.privateHandler(CommandParser.java:81) ~[BotV4.jar:?]
	at com.author404e.bot.util.EventHandlers.friend(EventHandlers.java:75) ~[BotV4.jar:?]
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method) ~[?:1.8.0_282]
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62) ~[?:1.8.0_282]
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43) ~[?:1.8.0_282]
	at java.lang.reflect.Method.invoke(Method.java:498) ~[?:1.8.0_282]
	at net.mamoe.mirai.internal.event.JvmMethodListenersInternalKt.registerEventHandler$callMethod$invokeWithErrorReport(JvmMethodListenersInternal.kt:140) ~[BotLibV4.jar:?]
	... 8 more
```


#### 版本及相关选项选择

- mirai-core: `2.7.0` <!--在 `` 中填写你正在使用的版本号，如 `2.6.7`-->
- bot-protocol: `ANDROID_PAD` <!-- 在 `` 中填写你正在使用的登录协议，如 `ANDROID_PHONE` -->

<!--如果你正在通过 mirai-console 或其他间接使用 mirai-core 的中间件（如 mirai-native、mirai-api-http 或者其他第三方 SDK）, 请同样提供版本号-->

--- a/mirai-core/src/commonMain/kotlin/network/notice/priv/PrivateMessageProcessor.kt
+++ b/mirai-core/src/commonMain/kotlin/network/notice/priv/PrivateMessageProcessor.kt
@@ -19,6 +19,7 @@ import net.mamoe.mirai.internal.network.components.SimpleNoticeProcessor
 import net.mamoe.mirai.internal.network.components.SsoProcessor
 import net.mamoe.mirai.internal.network.notice.group.GroupMessageProcessor
 import net.mamoe.mirai.internal.network.protocol.data.proto.MsgComm
+import net.mamoe.mirai.internal.network.protocol.packet.chat.voice.PttStore
 import net.mamoe.mirai.utils.assertUnreachable
 import net.mamoe.mirai.utils.context
 
@@ -48,6 +49,16 @@ internal class PrivateMessageProcessor : SimpleNoticeProcessor<MsgComm.Msg>(type
             166, 167, // 单向好友
             208, // friend ptt, maybe also support stranger
             -> {
+                data.msgBody.richText.ptt?.let { ptt ->
+                    if (ptt.downPara.isEmpty()) {
+                        val rsp = bot.network.sendAndExpect(
+                            PttStore.C2CPttDown(bot.client, senderUin, ptt.fileUuid)
+                        )
+                        if (rsp is PttStore.C2CPttDown.Response.Success) {
+                            ptt.downPara = rsp.downloadUrl.encodeToByteArray()
+                        }
+                    }
+                }
                 handlePrivateMessage(
                     data,
                     bot.getFriend(senderUin)?.impl()

--- a/mirai-core/src/commonMain/kotlin/network/protocol/data/proto/Msg.kt
+++ b/mirai-core/src/commonMain/kotlin/network/protocol/data/proto/Msg.kt
@@ -760,7 +760,7 @@ internal class ImMsgBody : ProtoBuf {
         @ProtoNumber(17) @JvmField val pttUrl: ByteArray = EMPTY_BYTE_ARRAY,
         @ProtoNumber(18) @JvmField val groupFileKey: ByteArray = EMPTY_BYTE_ARRAY,
         @ProtoNumber(19) @JvmField val time: Int = 0,
-        @ProtoNumber(20) @JvmField val downPara: ByteArray = EMPTY_BYTE_ARRAY,
+        @ProtoNumber(20) @JvmField var downPara: ByteArray = EMPTY_BYTE_ARRAY,
         @ProtoNumber(29) @JvmField val format: Int = 0,
         @ProtoNumber(30) @JvmField val pbReserve: ByteArray = EMPTY_BYTE_ARRAY,
         @ProtoNumber(31) @JvmField val bytesPttUrls: List<ByteArray> = emptyList(),

--- a/mirai-core/src/commonMain/kotlin/network/protocol/packet/PacketFactory.kt
+++ b/mirai-core/src/commonMain/kotlin/network/protocol/packet/PacketFactory.kt
@@ -142,6 +142,7 @@ internal object KnownPacketFactories {
         ImgStore.GroupPicUp,
         PttStore.GroupPttUp,
         PttStore.GroupPttDown,
+        PttStore.C2CPttDown,
         LongConn.OffPicUp,
 //        LongConn.OffPicDown,
         TroopManagement.EditSpecialTitle,

--- a/mirai-core/src/commonMain/kotlin/network/protocol/packet/chat/voice/PttStore.kt
+++ b/mirai-core/src/commonMain/kotlin/network/protocol/packet/chat/voice/PttStore.kt
@@ -195,6 +195,50 @@ internal class PttStore {
         }
     }
 
+    object C2CPttDown : OutgoingPacketFactory<C2CPttDown.Response>(
+        "PttCenterSvr.pb_pttCenter_CMD_REQ_APPLY_DOWNLOAD-1200"
+    ) {
+        operator fun invoke(client: QQAndroidClient, uin: Long, uuid: ByteArray) =
+            buildOutgoingUniPacket(client) {
+                writeProtoBuf(
+                    Cmd0x346.ReqBody.serializer(), Cmd0x346.ReqBody(
+                        msgApplyDownloadReq = Cmd0x346.ApplyDownloadReq(
+                            uin = uin,
+                            uuid = uuid,
+                            needHttpsUrl = 1,
+                        ),
+                        clientType = 104,
+                        cmd = 1200,
+                        businessId = 17, // or 3?
+                    )
+                )
+            }
+
+        sealed class Response : Packet {
+            class Failed(val retMsg: String) : Response() {
+                override fun toString(): String {
+                    return "PttCenterSvr.pb_pttCenter#download.Failed(retMsg=$retMsg)"
+                }
+            }
+
+            class Success(val downloadUrl: String) : Response() {
+                override fun toString(): String {
+                    return "PttCenterSvr.pb_pttCenter#download.Success"
+                }
+            }
+        }
+
+        override suspend fun ByteReadPacket.decode(bot: QQAndroidBot): C2CPttDown.Response {
+            val data = readProtoBuf(Cmd0x346.RspBody.serializer())
+            val rsp = data.msgApplyDownloadRsp ?: return Response.Failed("Response not found")
+            if (rsp.retMsg != "success") {
+                return Response.Failed(rsp.retMsg)
+            }
+            val downloadInfo = rsp.msgDownloadInfo ?: return Response.Failed("Download info not found")
+            return Response.Success(downloadInfo.downloadUrl)
+        }
+    }
+
     object C2C {
         fun createC2CPttStoreBDHExt(
             bot: QQAndroidBot,

